Feat/rand notify

hlongvu opened the pull request:
<!-- PLEASE FILL OUT THE FOLLOWING MARKDOWN TEMPLATE -->
<!-- PR title alone should be sufficient to understand changes. -->
Random value notification
## Change Description
<!-- Describe your changes, their justification, AND their impact. -->

## Documentation Additions
- [ ] Documentation Additions
<!-- checked [x] = Documentation changes; unchecked [ ] = no changes, ignore this section -->
<!-- Describe what must be added to the documentation after merge. -->

## Pushing The Code
<!-- Push all your changes as one commit. Rebase as needed. -->

## Unit Test Coverage
<!-- Provide unit test coverage that encompasses all new functionality. -->

hlongvu committed:
init notification logic

hlongvu committed:
update code hash logic

hlongvu committed:
init test

hlongvu committed:
test hash record

hlongvu committed:
build test contract

hlongvu committed:
test for allow list legacy callback

hlongvu committed:
add contract for test notification

hlongvu committed:
fix test

hlongvu committed:
test for code hash

porkchop commented on include/orng.hpp:
```diff
@@ -250,6 +308,18 @@ CONTRACT orng : public eosio::contract
+        eosio::name dapp;
+        eosio::checksum256 code_hash;
+        eosio::time_point_sec added_time;
+        eosio::time_point_sec sunset_time;
```
I know I asked for this, but now I have second thoughts that we should enforce any sunset time :/ I think it should be removed unfortunately

porkchop reviewed (changes requested):
As I said in another thread, I thik we can remove `reguser` now: https://github.com/worldwide-asset-exchange/wax-orng/pull/64/files#diff-37fc76b080dd1a51fa4cc675a47c6db193c7bb0106b07c52a61a4384e20184dbL125
To reintro the auto creation (by orng.wax RAM) of stake and acct entries you can look at my commit here to make it easier: https://github.com/worldwide-asset-exchange/wax-orng/pull/48/commits/a3c272c1d919743d876e88a7c8c48434d4c75740
Also, we should update the example code to use the new require_recipient methodology.

The PR is extremely well implemented despite my second thoughts

hlongvu committed:
remove reguser

hlongvu committed:
remove sunset time

hlongvu committed:
update dice game example

danielvo11 commented:
We should also update example code in [README.md](https://github.com/worldwide-asset-exchange/wax-orng/blob/feat/rand-notify/README.md#1-smart-contract-integration) it still using receiverand function.

danielvo11 commented:
LGTM

porkchop committed:
* only notify or callback not both
* give new dapps a way to use notify during collection phase: skiplegacy action
* Explain it all in the readme

porkchop committed:
no dual callback mode

porkchop committed:
v2.0->v3.0

porkchop committed:
fix makefile

porkchop committed:
fix makefile #2

porkchop committed:
typo

porkchop committed:
fix index in tests

porkchop committed:
fix example makefile

porkchop committed:
cleanup deprecated registration logic

porkchop committed:
Revise staking rates and maturity periods in README

Updated staking and payment details in README.md, including changes to rates, maturity periods, and example commands.
